### ISSUE TITLE
perf_analyzer --async miscommunicates with server and runs out of memory

### ISSUE THREAD

**Issue by raimondasl** [triton-inference-server/server#2686]
I have been running perf_analyzer --async and have seen the following issues.
As suggested by Triton team the server or perf_analyzer gets killed because of OOM.
I think that it runs OOM because of some client-side or server-side issue. Possibly different issue depending on the scenario below.
I am running `simple` model from Triton examples. I am using 21.02 Triton server and SDK for perf_analyzer.
Config file:

name: "simple"
platform: "tensorflow_graphdef"
max_batch_size: 10000
input [
  {
    name: "INPUT0"
    data_type: TYPE_INT32
    dims: [ 16 ]
  },
  {
    name: "INPUT1"
    data_type: TYPE_INT32
    dims: [ 16 ]
  }
]
output [
  {
    name: "OUTPUT0"
    data_type: TYPE_INT32
    dims: [ 16 ]
  },
  {
    name: "OUTPUT1"
    data_type: TYPE_INT32
    dims: [ 16 ]
  }
]
dynamic_batching {
    preferred_batch_size: [ 32 ]
    max_queue_delay_microseconds: 100
}
instance_group [
{
count: 2
kind: KIND_CPU
}
]
 
I am listing the behavior from worst towards acceptable.
I don't attach long server-side and perf_analyzer-side logs yet - not sure if this is possible here.
  
So now behavior from worst to acceptable:
 
Worst:
I saw tritonserver get killed when running perf_analyzer --async. I can't seem to replicate this though. I'll add a comment if I can replicate.
 
Next worst behavior:
This seems to happen when perf_analyzer has 1000 threads and request rate is 25K+ or so.
`perf_analyzer -m simple -i gRPC --async --max-threads 1000 --request-rate-range 25000 -p 40000 -u myip:8001`
Server shows some events in short log, then stops showing anything, while perf_analyzer continues to run, but is eventually killed due to OOM:
/workspace# perf_analyzer -m simple -i gRPC --async --max-threads 1000 --request-rate-range 25000 -p 40000 -u myip:8001
*** Measurement Settings ***
  Batch size: 1
  Measurement window: 40000 msec
  Using uniform distribution on request generation
  Using asynchronous calls for inference
  Stabilizing using average latency
 
Request Rate: 25000 inference requests per seconds
Killed

Server log:
... 
I0331 16:57:41.923224 289 http_server.cc:2717] Started HTTPService at 0.0.0.0:8000
I0331 16:57:41.965419 289 http_server.cc:2736] Started Metrics Service at 0.0.0.0:8002
I0331 16:58:00.133094 289 grpc_server.cc:270] Process for ModelMetadata, rpc_ok=1, 0 step START
I0331 16:58:00.133123 289 grpc_server.cc:225] Ready for RPC 'ModelMetadata', 1
I0331 16:58:00.133132 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.133143 289 model_repository_manager.cc:452] VersionStates() 'simple'
I0331 16:58:00.133236 289 grpc_server.cc:270] Process for ModelMetadata, rpc_ok=1, 0 step COMPLETE
I0331 16:58:00.133246 289 grpc_server.cc:408] Done for ModelMetadata, 0
I0331 16:58:00.134955 289 grpc_server.cc:270] Process for ModelConfig, rpc_ok=1, 0 step START
I0331 16:58:00.134980 289 grpc_server.cc:225] Ready for RPC 'ModelConfig', 1
I0331 16:58:00.134994 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.136328 289 grpc_server.cc:270] Process for ModelConfig, rpc_ok=1, 0 step COMPLETE
I0331 16:58:00.136364 289 grpc_server.cc:408] Done for ModelConfig, 0
I0331 16:58:00.137590 289 grpc_server.cc:270] Process for ServerMetadata, rpc_ok=1, 0 step START
I0331 16:58:00.137608 289 grpc_server.cc:225] Ready for RPC 'ServerMetadata', 1
I0331 16:58:00.137653 289 grpc_server.cc:270] Process for ServerMetadata, rpc_ok=1, 0 step COMPLETE
I0331 16:58:00.137663 289 grpc_server.cc:408] Done for ServerMetadata, 0
I0331 16:58:00.292145 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 1 step START
I0331 16:58:00.292164 289 grpc_server.cc:3082] New request handler for ModelInferHandler, 4
I0331 16:58:00.292173 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.292181 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.292197 289 infer_request.cc:502] prepared: [0x0x7f66680035a0] request id: 0, model: simple, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
original inputs:
[0x0x7f6668003928] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668003828] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
override inputs:
inputs:
[0x0x7f6668003828] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668003928] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
original requested outputs:
OUTPUT0
OUTPUT1
requested outputs:
OUTPUT0
OUTPUT1
 
I0331 16:58:00.292416 289 tensorflow.cc:2097] model simple, instance simple_0, executing 1 requests
I0331 16:58:00.292430 289 tensorflow.cc:1386] TRITONBACKEND_ModelExecute: Running simple_0 with 1 requests
I0331 16:58:00.292571 289 tensorflow.cc:1614] TRITONBACKEND_ModelExecute: input 'INPUT0' is GPU tensor: false
I0331 16:58:00.292582 289 tensorflow.cc:1614] TRITONBACKEND_ModelExecute: input 'INPUT1' is GPU tensor: false
I0331 16:58:00.295056 289 infer_response.cc:165] add response output: output: OUTPUT0, type: INT32, shape: [1,16]
I0331 16:58:00.295073 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT0', size: 64, addr: 0x7f66a4004380
I0331 16:58:00.295083 289 tensorflow.cc:1797] TRITONBACKEND_ModelExecute: output 'OUTPUT0' is GPU tensor: false
I0331 16:58:00.295090 289 infer_response.cc:165] add response output: output: OUTPUT1, type: INT32, shape: [1,16]
I0331 16:58:00.295094 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT1', size: 64, addr: 0x7f66a40043d0
I0331 16:58:00.295101 289 tensorflow.cc:1797] TRITONBACKEND_ModelExecute: output 'OUTPUT1' is GPU tensor: false
I0331 16:58:00.295106 289 grpc_server.cc:3240] ModelInferHandler::InferResponseComplete, 1 step ISSUED
I0331 16:58:00.295118 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a4004380
I0331 16:58:00.295125 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a40043d0
I0331 16:58:00.295180 289 grpc_server.cc:2817] ModelInferHandler::InferRequestComplete
I0331 16:58:00.295189 289 tensorflow.cc:1855] TRITONBACKEND_ModelExecute: model simple_0 released 1 requests
I0331 16:58:00.295195 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 1 step COMPLETE
I0331 16:58:00.295207 289 grpc_server.cc:2139] Done for ModelInferHandler, 1
I0331 16:58:00.304772 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 4 step START
I0331 16:58:00.304787 289 grpc_server.cc:3082] New request handler for ModelInferHandler, 5
I0331 16:58:00.304798 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304813 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304831 289 infer_request.cc:502] prepared: [0x0x7f6668004d10] request id: 0, model: simple, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
original inputs:
[0x0x7f6668005058] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668004f98] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
override inputs:
inputs:
[0x0x7f6668004f98] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668005058] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
original requested outputs:
OUTPUT0
OUTPUT1
requested outputs:
OUTPUT0
OUTPUT1
 
I0331 16:58:00.304855 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 5 step START
I0331 16:58:00.304862 289 grpc_server.cc:3082] New request handler for ModelInferHandler, 6
I0331 16:58:00.304870 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304874 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304885 289 infer_request.cc:502] prepared: [0x0x7f6668005ea0] request id: 0, model: simple, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
original inputs:
[0x0x7f6668006228] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668006128] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
override inputs:
inputs:
[0x0x7f6668006128] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668006228] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
original requested outputs:
OUTPUT0
OUTPUT1
requested outputs:
OUTPUT0
OUTPUT1
 
I0331 16:58:00.304940 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 6 step START
I0331 16:58:00.304955 289 grpc_server.cc:3082] New request handler for ModelInferHandler, 7
I0331 16:58:00.304962 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304967 289 model_repository_manager.cc:496] GetInferenceBackend() 'simple' version -1
I0331 16:58:00.304980 289 infer_request.cc:502] prepared: [0x0x7f6668007d00] request id: 0, model: simple, requested version: -1, actual version: 1, flags: 0x0, correlation id: 0, batch size: 1, priority: 0, timeout (us): 0
original inputs:
[0x0x7f6668008088] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668007f88] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
override inputs:
inputs:
[0x0x7f6668007f88] input: INPUT0, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
[0x0x7f6668008088] input: INPUT1, type: INT32, original shape: [1,16], batch + shape: [1,16], shape: [16]
original requested outputs:
OUTPUT0
OUTPUT1
requested outputs:
OUTPUT0
OUTPUT1
 
I0331 16:58:00.305020 289 tensorflow.cc:2097] model simple, instance simple_0, executing 3 requests
I0331 16:58:00.305032 289 tensorflow.cc:1386] TRITONBACKEND_ModelExecute: Running simple_0 with 3 requests
I0331 16:58:00.305044 289 tensorflow.cc:1614] TRITONBACKEND_ModelExecute: input 'INPUT0' is GPU tensor: false
I0331 16:58:00.305052 289 tensorflow.cc:1614] TRITONBACKEND_ModelExecute: input 'INPUT1' is GPU tensor: false
I0331 16:58:00.305152 289 infer_response.cc:165] add response output: output: OUTPUT0, type: INT32, shape: [1,16]
I0331 16:58:00.305165 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT0', size: 64, addr: 0x7f66a400dd90
I0331 16:58:00.305170 289 infer_response.cc:165] add response output: output: OUTPUT0, type: INT32, shape: [1,16]
I0331 16:58:00.305175 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT0', size: 64, addr: 0x7f66a4004380
I0331 16:58:00.305178 289 infer_response.cc:165] add response output: output: OUTPUT0, type: INT32, shape: [1,16]
I0331 16:58:00.305182 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT0', size: 64, addr: 0x7f66a4011810
I0331 16:58:00.305186 289 tensorflow.cc:1797] TRITONBACKEND_ModelExecute: output 'OUTPUT0' is GPU tensor: false
I0331 16:58:00.305190 289 infer_response.cc:165] add response output: output: OUTPUT1, type: INT32, shape: [1,16]
I0331 16:58:00.305194 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT1', size: 64, addr: 0x7f66a40201d0
I0331 16:58:00.305198 289 infer_response.cc:165] add response output: output: OUTPUT1, type: INT32, shape: [1,16]
I0331 16:58:00.305201 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT1', size: 64, addr: 0x7f66a40043d0
I0331 16:58:00.305205 289 infer_response.cc:165] add response output: output: OUTPUT1, type: INT32, shape: [1,16]
I0331 16:58:00.305209 289 grpc_server.cc:2230] GRPC: using buffer for 'OUTPUT1', size: 64, addr: 0x7f66a400d970
I0331 16:58:00.305212 289 tensorflow.cc:1797] TRITONBACKEND_ModelExecute: output 'OUTPUT1' is GPU tensor: false
I0331 16:58:00.305216 289 grpc_server.cc:3240] ModelInferHandler::InferResponseComplete, 4 step ISSUED
I0331 16:58:00.305222 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a400dd90
I0331 16:58:00.305225 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a40201d0
I0331 16:58:00.305263 289 grpc_server.cc:3240] ModelInferHandler::InferResponseComplete, 5 step ISSUED
I0331 16:58:00.305269 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a4004380
I0331 16:58:00.305272 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a40043d0
I0331 16:58:00.305281 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 4 step COMPLETE
I0331 16:58:00.305300 289 grpc_server.cc:2139] Done for ModelInferHandler, 4
I0331 16:58:00.305288 289 grpc_server.cc:3240] ModelInferHandler::InferResponseComplete, 6 step ISSUED
I0331 16:58:00.305310 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 5 step COMPLETE
I0331 16:58:00.305330 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a4011810
I0331 16:58:00.305338 289 grpc_server.cc:2265] GRPC free: size 64, addr 0x7f66a400d970
I0331 16:58:00.305348 289 grpc_server.cc:2139] Done for ModelInferHandler, 5
I0331 16:58:00.305389 289 grpc_server.cc:2817] ModelInferHandler::InferRequestComplete
I0331 16:58:00.305399 289 grpc_server.cc:2817] ModelInferHandler::InferRequestComplete
I0331 16:58:00.305408 289 grpc_server.cc:2817] ModelInferHandler::InferRequestComplete
I0331 16:58:00.305411 289 grpc_server.cc:3089] Process for ModelInferHandler, rpc_ok=1, 6 step COMPLETE
I0331 16:58:00.305416 289 tensorflow.cc:1855] TRITONBACKEND_ModelExecute: model simple_0 released 3 requests
I0331 16:58:00.305431 289 grpc_server.cc:2139] Done for ModelInferHandler, 6
 
Next worst:
 
This seems to happen when perf_analyzer has 100 threads and request rate 10K-25K or so.
`perf_analyzer -m simple -i gRPC --async --max-threads 100 --request-rate-range 25000 -p 40000 -u myip:8001`
Server processes a lot of requests (unlike the above where it processes only couple), but then stops as if all requests have been processed (also unlike the above).
perf_analyzer keeps running after server is not doing anything anymore. Eventually perf_analyzer gets killed due to OOM.
 
/workspace# perf_analyzer -m simple -i gRPC --async --max-threads 100 --request-rate-range 25000 -p 40000 -u myip:8001
*** Measurement Settings ***
  Batch size: 1
  Measurement window: 40000 msec
  Using uniform distribution on request generation
  Using asynchronous calls for inference
  Stabilizing using average latency
 
Request Rate: 25000 inference requests per seconds
Killed
 
Next:
 
This seems to happen when perf_analyzer has 100 threads and the server load is lower (5K requests in this case).
perf_analyzer eventually prints " No valid requests recorded within time interval. Please use a larger time window."
Server continues processing something after this message is printed. Eventually (after couple minutes) server stops processing stuff and perf_analyzer exits.
 
/workspace# perf_analyzer -m simple -i gRPC --async --max-threads 100 --request-rate-range 5000 -p 40000 -u myip:8001
*** Measurement Settings ***
  Batch size: 1
  Measurement window: 40000 msec
  Using uniform distribution on request generation
  Using asynchronous calls for inference
  Stabilizing using average latency
 
Request Rate: 5000 inference requests per seconds
No valid requests recorded within time interval. Please use a larger time window.
 
Finally, successful run: 100 threads, 1K requests per second:
perf_analyzer --sync works fine with up to 1000 threads/37K requests per second without delayed requests. perf_analyzer --sync does not fail/error even with higher load. perf_analyzer --async is way worse in terms of what it can handle.
 
------------------------------------------------------------
/workspace# perf_analyzer -m simple -i gRPC --async --max-threads 100 --request-rate-range 1000 -p 40000 -u myip:8001
*** Measurement Settings ***
  Batch size: 1
  Measurement window: 40000 msec
  Using uniform distribution on request generation
  Using asynchronous calls for inference
  Stabilizing using average latency
 
Request Rate: 1000 inference requests per seconds
  Client:
    Request count: 40000
    Throughput: 1000 infer/sec
    Avg latency: 752 usec (standard deviation 149 usec)
    p50 latency: 740 usec
    p90 latency: 784 usec
    p95 latency: 807 usec
    p99 latency: 1059 usec
    Avg gRPC time: 745 usec ((un)marshal request/response 4 usec + response wait 741 usec)
  Server:
    Inference count: 48000
    Execution count: 47985
    Successful request count: 47985
    Avg request latency: 354 usec (overhead 51 usec + queue 176 usec + compute input 20 usec + compute infer 66 usec + compute output 41 usec)
 
Inferences/Second vs. Client Average Batch Latency
Request Rate: 1000, throughput: 1000 infer/sec, latency 752 usec

**Comment by raimondasl** [comment]
As discussed, I am adding my additional experiments:

Results with 1 client machine

perf_analyzer with both --sync and --async works fine up to 35K requests/s when Azure H16 VM is used to run perf_analyzer client. (35K requests per second is 1K threads sending requests 35 times/second per thread).

                 Avg     p90     p95     p99
--sync	1.8ms	2.9ms	4.8ms	12.4ms
--async	.9ms	1ms	    1.5ms	5.3ms

With higher number of requests/s --sync degrades latency a lot above real-time. --async starts failing (sometimes it fails even at 35K) likely due to some issue of threading or miscommunication with server.

Results with 2 client machines

perf_analyzer with both --sync and --async works fine up to 50K requests/s when two Azure H16 VMs are used to run two perf_analyzer clients. Each perf_analyzer client runs 500 threads. (50K requests per second is 1K threads sending requests 50 times/second per thread).

                 Avg     p90     p95     p99
--sync	1.3ms	1.7ms	2.9ms	6ms
--async	1.2ms	1.1ms	1.1ms	5.5ms

I succeeded running --async with 60K req/s (30K req/s from each of 2 clients, each client running 500 threads). Avg latency 1.1ms, p90 latency 1.2ms.

I think this shows that the --async issue is on the client side and not on the server side. It looks like the client program (perf_analyzer) has issues when it has to run a lot of threads and each thread has to run a lot of requests.

Also, if I run perf_analyzer on Azure D16as_v4 VM that has worse CPU/network than H16 VM, the --async seems to die at much lower req/s. The following dies on Azure D16as_v4 VM, while it works fine on Azure H16 VM with the same server setup:
` perf_analyzer -m simple -i gRPC --async --max-threads 1000 --request-rate-range 20000 -p 40000 -u myip:8001`

Hope this helps. Feel free to contact me directly and I can walk through the things I run.

**Comment by GuanLuo** [comment]
> --async starts failing

Do you mean the perf_analyzer get killed by the system?

**Comment by raimondasl** [comment]
@GuanLuo : From what I see from perf_analyzer verbose output and server --log-verbose, at some point server stops receiving and processing requests from the client, while client thinks that it is still sending requests to the server ("Sent request" log lines). From that point the client memory consumption starts rising until it runs OOM and gets killed by the system. I think I tried to show that in the log fragments above. I am going to try to attach log files here although without seeing printouts in real time, they are harder to interpret.
Server log:
[server.log](https://github.com/triton-inference-server/server/files/6408577/server.log)
perf_analyzer log was too long to attach. I removed 4.5M lines from the middle part of it which is pretty much the same content:

[perf_analyzer_snipped.log](https://github.com/triton-inference-server/server/files/6408625/perf_analyzer_snipped.log)

Two notes:
- server log has trailing output after I ctrl-c'd the server
- as mentioned above perf_analyzer client is writing the "Sent request ..." for a long time after the server is no longer receiving them. This is where it's consuming more and more memory until it OOMs and gets killed.

**Comment by GuanLuo** [comment]
With `--request-rate-range`, the perf_analyzer will send the request in the given rate regardless, I think the OOM is due to the server is overloaded and causing a stack of pending request (/ request handle) on both the client and server. In such case, I think some kind of queue policy should be set to reject requests if the queue size exceeds a predefined size or the requests have been waiting in the queue for too long.

Another thing is that I think the server is still capable of receiving and processing requests after processing the high load. From my local experiment I logged the number of requests queued in the server, and I observed that the number was increasing in general. Once the client is killed due to OOM (stop generating new requests), the number gradually decreased which suggested that the server is still processing the request. And sending another request after the server processed all previous request still returns proper response.

**Comment by raimondasl** [comment]
@GuanLuo : I have explained upthread why this is not due to server overload. The server is not loaded at all when it happens. What you describe in second paragraph is not what I see.

**Comment by raimondasl** [comment]
@GuanLuo : Yes, that's what I believe is happening. And you asked a great question. I just tested it with Triton 21.05: 2 clients running on 2 different machines. Server running on 3rd machine. Started server with --log-verbose 1 . Started 1st client on Azure D16as_v4 VM with `perf_analyzer -m simple -i gRPC --async --max-threads 1000 --request-rate-range 30000 -p 40000 -u myip:8001` . Waited until no activity shows on server screen log. The 1st client is still "running" and its memory is growing, the server shows no activity. Then I started 2nd client with `perf_analyzer -m simple -i gRPC --async --max-threads 10 --request-rate-range 100 -p 40000 -u myip:8001`. Second client ran fine and server showed activity: requests being processed. Stopped 2nd client. 1st client is still "running" but no activity is showing on server.

I believe this shows that it is a client-side issue. The server is fine - as I thought upthread. 

This is somewhat difficult to describe in writing and difficult to attach relevant log portions. If you would like to see my setup and see the failure directly, I'd be happy to show you what's happening. You can get my contact details from @deadeyegoodwin and we can schedule a meeting. 

Thanks for looking into this.

**Comment by GuanLuo** [comment]
Some update on this issue:

I am able to reproduce the behavior that you have seen, and I have the new finding that shows that the issue may be due to the GRPC (or the use of GRPC) on the client side:

When I observed that the server was not receiving requests, instead of let perf_analyzer running and eventually got killed, I used "ctrl + c" which caused perf_analzyer to exit early gracefully, then perf_analyzer stopped generating new requests and waited for the completion of the previous requests as expected. And I saw that the server started receiving requests, which I believe is the requests that perf_analyzer generated until receiving the exit signal (number reported from server matched the number of "sent request").

It appears that the request are queued in GRPC's client stub but not being flushed until perf_analyzer receives a signal, we will look into it more and see how to resolve this.
